## DATA 690 Homework 2 (50 points - Due on Tuesday, September 26, 2023 by 11:59 pm ET)

The output of this assignment for submission should be in PDF format **AND** .py or .ipynb.

The name of the file should be as follows:

    Lastname_Firstname_Homework3.pdf (example: Thomas_Sunela_Homework3.pdf)
   **AND**
   
    Lastname_Firstname_Homework3.ipynb (example: Thomas_Sunela_Homework3.ipynb.
    
In short, you are submitting the python notebook as well as the pdf of that notebook.
Do **NOT** submit .html file, the system will give you an error.

    
**Incorrect file name will cost you points!<br>**

Instructions for converting a Jupyter Python notebook to PDF:
                                        
    Go to the menu and choose, File --> Download As --> html.
    Open that html file and print it to PDF.
    Submit the PDF file **NOT** the html file.

If you are using Google Colab, remember to review the PDF before submitting to ensure that all cells and answers are displayed in the PDF.


**Things to note:**

* Each cell should display an output
* Use both Markdown and code comments in the Jupyter Notebook as needed

### IF YOU ARE MAKING ANY ASSUMPTIONS, WRITE THAT IN A MARKDOWN CELL OR COMMENT
### $\color{red}{\text{Answer the questions asked as well, not just code}}$



We will be using the SF Salaries dataset from Kaggle! The dataset is provided to you in Blackboard.  

**#1 Import pandas as pd**

In [2]:
import pandas as pd

**#2 Read Salaries.csv as a dataframe called sal**

In [12]:
from google.colab import files


uploaded = files.upload()


import io

sal = pd.read_csv("Salaries.csv")

Saving Salaries.csv to Salaries (1).csv


<ipython-input-12-7a871c0bbebc>:9: DtypeWarning: Columns (12) have mixed types. Specify dtype option on import or set low_memory=False.
  sal = pd.read_csv("Salaries.csv")


\**#3 Check the head of the DataFrame. Is there anything unique? What can you tell about the data?** (5 points)

In [ ]:
sal.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.00  400184.25       NaN  567595.43         567595.43   
1  155966.02    245131.88  137811.38       NaN  538909.28         538909.28   
2  212739.13    106088.18   16452.60       NaN  335279.91         335279.91   
3   77916.00     56120.71  198306.90       NaN  332343.61         332343.61   
4  134401.60      9737.00  182234.59       NaN  326373.19         326373.19   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN  
3  2011    NaN  San Francisco    NaN  
4  2011    NaN  San Francisco    NaN

Written Answer: The only employee who doesn't have overtime pay is Nathaniel Ford, who is the General Manager of the MTA. There are two employees who are both in the Police Departments and both hold "Captain III" title. There are not any values in the "Benefits" column which is interesting. All of the employees' agencies are based in San Fransisco which could mean that benefits are not provided to working employees employed in the city or state of California.

**#4 Use the .info() method to find out how many entries there are. Can you tell anything more about the data?** (3 points)

In [ ]:
sal.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148650 entries, 0 to 148649
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148650 non-null  int64  
 1   EmployeeName      148650 non-null  object 
 2   JobTitle          148650 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148650 non-null  float64
 8   TotalPayBenefits  148650 non-null  float64
 9   Year              148650 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148650 non-null  object 
 12  Status            38119 non-null   object 
dtypes: float64(7), int64(2), object(4)
memory usage: 14.7+ MB


Written Answer: There are three different data types: "float64", "int64" and "object". Majority of the data are floats. It takes up 14.7+ MB of memory.

**#5 What is the average BasePay?** (3 points)

In [ ]:
#calculate the mean
sal['BasePay'].mean()

66325.4488404877

**#6 What is the highest amount of OvertimePay in the dataset?** (3 points)

In [ ]:
#find the max value in OverTime Pay
sal['OvertimePay'].max()

245131.88

**#7 What is the job title of MARTY ROSS?** (4 points)

In [ ]:
sal[sal['EmployeeName'] == 'MARTY ROSS' ] ['JobTitle']

18    BATTALION CHIEF, (FIRE DEPARTMENT)
Name: JobTitle, dtype: object

**#8 How much does JOSEPH DRISCOLL make (including benefits)?** (4 points)

In [ ]:
sal [ sal['EmployeeName'] == 'JOSEPH DRISCOLL' ] ['TotalPayBenefits']

24    270324.91
Name: TotalPayBenefits, dtype: float64

**#9 What is the name of highest paid person (including benefits)?** (3 points)

In [4]:
sal.iloc[sal['TotalPayBenefits'].idxmax()]

Id                                                               1
EmployeeName                                        NATHANIEL FORD
JobTitle            GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY
BasePay                                                  167411.18
OvertimePay                                                    0.0
OtherPay                                                 400184.25
Benefits                                                       NaN
TotalPay                                                 567595.43
TotalPayBenefits                                         567595.43
Year                                                          2011
Notes                                                          NaN
Agency                                               San Francisco
Status                                                         NaN
Name: 0, dtype: object

**#10 What is the name of lowest paid person (including benefits)? Do you notice something strange about how much he or she is paid?** (5 points)

In [5]:
sal.iloc[sal['TotalPayBenefits'].idxmin()]

Id                                      148654
EmployeeName                         Joe Lopez
JobTitle            Counselor, Log Cabin Ranch
BasePay                                    0.0
OvertimePay                                0.0
OtherPay                               -618.13
Benefits                                   0.0
TotalPay                               -618.13
TotalPayBenefits                       -618.13
Year                                      2014
Notes                                      NaN
Agency                           San Francisco
Status                                      PT
Name: 148649, dtype: object

Written Answer: The lowest paid person's name is Joe Lopez. It seems that he doesn't have a base pay and doesn't get paid overtime or benefits. His total pay is a negatice number which is very strange.

**#11 What was the average (mean) BasePay of all employees in 2012?** (5 points)

In [14]:
sal.groupby('Year').mean()['BasePay']

<ipython-input-14-a42aa5aba483>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  sal.groupby('Year').mean()['BasePay']


Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

**#12 How many unique job titles are there?** (5 points)

In [ ]:
sal['JobTitle'].nunique()

2158

**#13 What are the top 5 most common jobs?** (5 points)

In [ ]:
sal['JobTitle'].value_counts().head()

Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64

**#14 How many Job Titles were represented by only one person in 2012? (e.g. Job Titles with only one occurence in 2012?)** (5 points)

In [ ]:
sum(sal[sal['Year']==2012]['JobTitle'].value_counts() == 1)

190